File: backend/app/api/release.py
from fastapi import APIRouter, Depends

from app.api.deps import get_runtime
from app.schemas.release import CleanroomVerificationRecord, ReleasePackageRecord, ReleaseStatus
from app.services.runtime_state import RuntimeState

router = APIRouter(prefix="/api/release", tags=["release"])


@router.get("/status", response_model=ReleaseStatus)
def status(runtime: RuntimeState = Depends(get_runtime)) -> ReleaseStatus:
    return runtime.release.status(runtime)


@router.get("/preflight", response_model=ReleaseStatus)
def preflight(runtime: RuntimeState = Depends(get_runtime)) -> ReleaseStatus:
    return runtime.release.preflight(runtime)


@router.post("/check", response_model=ReleaseStatus)
def check(runtime: RuntimeState = Depends(get_runtime)) -> ReleaseStatus:
    return runtime.release.check(runtime)


@router.get("/cold-start-check", response_model=ReleaseStatus)
def cold_start_check(runtime: RuntimeState = Depends(get_runtime)) -> ReleaseStatus:
    return runtime.release.cold_start_check(runtime)


@router.post("/cold-start-check", response_model=ReleaseStatus)
def cold_start_check_post(runtime: RuntimeState = Depends(get_runtime)) -> ReleaseStatus:
    return runtime.release.cold_start_check(runtime)


@router.get("/check", response_model=ReleaseStatus)
def check_get(runtime: RuntimeState = Depends(get_runtime)) -> ReleaseStatus:
    return runtime.release.preflight(runtime)


@router.get("/package/latest", response_model=ReleasePackageRecord | None)
def latest_package(runtime: RuntimeState = Depends(get_runtime)) -> ReleasePackageRecord | None:
    return runtime.release.latest_package()


@router.post("/package/build", response_model=ReleasePackageRecord)
def build_package(runtime: RuntimeState = Depends(get_runtime)) -> ReleasePackageRecord:
    return runtime.release.build_package(runtime)


@router.get("/clean-room/latest", response_model=CleanroomVerificationRecord | None)
def latest_cleanroom(runtime: RuntimeState = Depends(get_runtime)) -> CleanroomVerificationRecord | None:
    return runtime.release.latest_cleanroom_verification()


@router.post("/clean-room/run", response_model=CleanroomVerificationRecord)
def run_cleanroom(runtime: RuntimeState = Depends(get_runtime)) -> CleanroomVerificationRecord:
    return runtime.release.run_cleanroom_verification(runtime)

File: frontend/src/api/release.ts
import type { CleanroomVerificationRecord, ReleasePackageRecord, ReleaseStatus } from '../types/release'

function apiBaseUrl(): string {
  const configured = import.meta.env.VITE_BACKEND_API_URL as string | undefined
  if (configured) return configured.replace(/\/$/, '')
  if (window.location.port && window.location.port !== '5173') return window.location.origin
  return `${window.location.protocol}//${window.location.hostname}:8000`
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBaseUrl()}${path}`, {
    headers: { 'Content-Type': 'application/json', ...init?.headers },
    ...init,
  })
  const body = await response.json()
  if (!response.ok) throw new Error(JSON.stringify(body))
  return body as T
}

export const fetchReleaseStatus = () => request<ReleaseStatus>('/api/release/status')
export const fetchReleasePreflight = () => request<ReleaseStatus>('/api/release/preflight')
export const runReleaseCheck = () => request<ReleaseStatus>('/api/release/check', { method: 'POST' })
export const runColdStartCheck = () => request<ReleaseStatus>('/api/release/cold-start-check')
export const fetchLatestReleasePackage = () => request<ReleasePackageRecord | null>('/api/release/package/latest')
export const buildReleasePackage = () => request<ReleasePackageRecord>('/api/release/package/build', { method: 'POST' })
export const fetchLatestCleanroom = () => request<CleanroomVerificationRecord | null>('/api/release/clean-room/latest')
export const runCleanroom = () => request<CleanroomVerificationRecord>('/api/release/clean-room/run', { method: 'POST' })
